Feature/chunk vcf dumper - EVA-301

pabarcgar opened the pull request.

pabarcgar committed:
VariantExporterTest simplified

pabarcgar committed:
VariantExporter: added new method to export just the variants in one region. Adapter tests to use that method

pabarcgar committed:
Added new method to RegionDivider

pabarcgar committed:
New RegionFactory class; Methods from RegionDivider moved to RegionFactory

pabarcgar committed:
Big refactoring: most of the vcf dumper moved from VariantExporter to VariantExporterController

pabarcgar committed:
Merged upstream develop

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporter.java:
```diff
@@ -87,227 +62,26 @@
-        while (variantDBIterator.hasNext()) {
-            Variant variant = variantDBIterator.next();
+        while (iterator.hasNext()) {
+            Variant variant = iterator.next();
```
These 2 lines could be converted into a for loop to match the style of the rest of iterators in the code.

pabarcgar replied:
To use a "for each" loop we need a class that implements "Iterable", and VariantDBIterator doesn't. VariantMongoDBAdaptor is "iterable", but if we use the for each loop over it, it will use the iterator() method with no parameters, that ignores the filters and options from QueryOptions. 

We can change this for a traditional for method: 
for (Variant v; v=variantIterator.next(); variantIterator.hasNext())

Do I change this?

cyenyxe replied:
Whatever you prefer. I thought VariantDBIterator would be supported, but if not, then I don't have any preference.

pabarcgar replied:
I won't change it then

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporter.java:
```diff
@@ -316,69 +90,24 @@ private int getVariantStart(QueryOptions query) {
+    public Map<String, VCFHeader> getVcfHeaders(Map<String, VariantSource> sources) throws IOException {
         Map<String, VCFHeader> headers = new TreeMap<>();
 
         for (VariantSource source : sources.values()) {
```
I checked the PASS metadata description in the 1kGP3 dataset and the description string is "All filters passed". I suggest using that one because that makes clearer the meaning of "PASS".

pabarcgar replied:
Changed in commit 8457291bd786c7b823fce40c7398dd7f23cdb323

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporter.java:
```diff
@@ -493,6 +189,7 @@ public int getFailedVariants() {
         Integer start = variant.getStart();
         Integer end = start + reference.length() -1;
         String filter = "PASS";
+        // TODO: avoid creating a list per variant
```
The way the list is used in this method, an array could be initialized as `new String[] { reference, alternate }`, the iterators would work the same and the get/set calls could be replaced by direct index access.

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporterController.java:
```diff
@@ -67,21 +94,142 @@ public QueryOptions getQuery(MultivaluedMap<String, String> queryParameters) {
+        Map<String, VariantSource> sources = exporter.getSources(variantSourceDBAdaptor, studies);
+        Map<String, VCFHeader> headers = exporter.getVcfHeaders(sources);
+        checkHeaders(headers);
+        Map<String, VariantContextWriter> writers = getWritters(headers);
```
Typo in "writters"

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporterController.java:
```diff
@@ -67,21 +94,142 @@ public QueryOptions getQuery(MultivaluedMap<String, String> queryParameters) {
-        return new VariantExporter(cellBaseClient, variantDBAdaptor, query).vcfExport(outputDir, variantSourceDBAdaptor);
+//    public List<String> run() throws URISyntaxException, IOException {
+//        return new VariantExporter(cellBaseClient, variantDBAdaptor, query).vcfExport(outputDir, variantSourceDBAdaptor);
+//    }
```
Remove commented block.

cyenyxe commented on eva-tools/src/test/java/embl/ebi/variation/eva/vcfDump/VariantExporterControllerTest.java:
```diff
@@ -0,0 +1,331 @@
+        List<String> studies = Arrays.asList("s1", "s2");
+        List<String> files = Arrays.asList("f3", "f4", "f5");
+
+        VariantExporterController controller = new VariantExporterController("hsapiens", null, studies, files, null, emptyFilter);
```
The constructor should check for parameters correctness, for instance, in a real use case it doesn't make sense that the database name is null.

cyenyxe commented on eva-tools/src/test/java/embl/ebi/variation/eva/vcfdump/VariantExporterTest.java:
```diff
@@ -84,190 +88,202 @@ public static void setUpClass() throws IOException, InterruptedException, URISyn
-        VariantSourceDBAdaptor variantSourceDBAdaptor = variantDBAdaptor.getVariantSourceDBAdaptor();
-
-        VariantExporter variantExporter = new VariantExporter(cellBaseClient, variantDBAdaptor, query);
+//    @Test
```
Remove commented code block, all the test methods seem to have already been migrated to VariantExporterControllerTest.

pabarcgar committed:
Removed old commented test methods

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporterController.java:
```diff
@@ -67,21 +94,142 @@ public QueryOptions getQuery(MultivaluedMap<String, String> queryParameters) {
+//                // TODO: log error
+//                sequenceDictionary = null;
+//            }
+            SAMSequenceDictionary sequenceDictionary = headers.get(studyId).getSequenceDictionary();
```
[According to the Javadoc](https://github.com/samtools/htsjdk/blob/8485225f488ac25b508f5876472d55a7844b5af7/src/main/java/htsjdk/variant/vcf/VCFHeader.java#L196-L197), this throws a SAMException if one or more contig lines do not have length information, which is something that could happen quite easily. @jmmut, can you comment on this?

jmmut replied:
Honestly, I don't remember what is a contig line. When I developed the first version of this, I had quite trouble with the docs, trying to understand what that dictionary was, and now I am in no better position to speak about this. :(

pabarcgar replied:
Example of contig line (human GRCh37 chromosome Y): 

#contig=<ID=Y,length=59373566,assembly=GCA_000001405.1,accession=CM000686.1>

Like incorrect contig lines are common, and contig lines are not mandatory in the VCF specification, then we should catch the exception to avoid the VCF exporter finishing with errors.

cyenyxe replied:
Sounds like a sensible solution :) I'm not even sure it's worth notifying the user...

pabarcgar committed:
Changed description of 'PASS' filter in vcf header

pabarcgar committed:
VariantExporter: changed List by Array

pabarcgar committed:
Added biodata dependency to avoid an incorrect biodata version being imported

pabarcgar committed:
Added log messages to VariantExportBootApplication

pabarcgar committed:
VariantExportController:
        - Refactor:
                New getVcfHeadersMethod
        - Catched SAMException when getting SequenceDictionary from VCF header
        - Fixed typo (writters)
        - Added some log messages

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExportBootApplication.java:
```diff
@@ -38,7 +39,7 @@
 public class VariantExportBootApplication implements CommandLineRunner {
 
-    final Logger logger = Logger.getLogger(getClass().getName());
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(VariantExportBootApplication.class);
```
Why not using the built-in logging library?

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporterController.java:
```diff
@@ -107,21 +108,34 @@ public QueryOptions getQuery(MultivaluedMap<String, String> queryParameters) {
-    public void run() throws IOException {
+    public void run() {
+        Map<String, VCFHeader> headers = getVcfHeaders();
+        if (headers != null) {
```
I think an Exception should be thrown if `headers` is null (or let it be directly thrown from `getVcfHeaders()`. The application won't work properly without them.

cyenyxe commented on eva-tools/src/main/java/embl/ebi/variation/eva/vcfdump/VariantExporterController.java:
```diff
@@ -170,16 +184,15 @@ private void checkHeaders(Map<String, VCFHeader> headers) {
+                sequenceDictionary = headers.get(studyId).getSequenceDictionary();
+            } catch (SAMException e) {
+                logger.warn("Incorrect sequence dictionary in study {}: {}", studyId, e.getMessage());
+                logger.warn("Sequence dictionary won't be included in study {} output VCF header", studyId);
```
Not sure people will understand what a "sequence dictionary" means XD I would suggest something less implementation-specific like:

> Incorrect sequence / contig meta-data found in study {}: {}
> It won't be included in study {} output VCF header

This they will know for sure what it is.

pabarcgar replied:
Changed in commit ece1f64667d4522a67e48825d42eef3e7e02f113

pabarcgar committed:
Changed some log messages

pabarcgar committed:
Removed commented code

pabarcgar committed:
VariantExporterController: dbName and output dir now cannot be null. Removed some TODOs

pabarcgar committed:
- VariantExportController: Check null arguments
